labgrid/util/marker.py

```diff
@@ -1,6 +1,187 @@
+        # character of the output buffer can't possibly be part of the output
+        # marker, it will be consumed.
+        partials = []
+        p = ""
+        for m in marker[:-1]:
+            p = p + m
+            partials.append(r"{}$".format(p))
+
+        # Add the complete marker, which doesn't need to be anchored to the end
+        # of the string to prevent a character from being consumed.
+        partials.append(marker)
+        partials.reverse()
+
+        self._output_re = re.compile(
+            r"^{}.".format("".join(r"(?!{})".format(p) for p in partials)).encode(
+                encoding
+            ),
+            re.DOTALL,
+        )
+
+        self._prompt_re = re.compile(prompt.encode(encoding))
+
+        self._eof_re = re.compile(
+            r"^{marker}\s+(\d+)\s+.*{prompt}".format(
+                marker=marker, prompt=prompt
+            ).encode(encoding)
+        )
+
+        self._expect = ConsoleProcessExpect(self, timeout=timeout)
+
+    def _handle_eof(self, match):
+        self.exitcode = int(match.group(1))
+        self._alive = False
+
+    @step(args=['size', 'timeout'], result=True)
+    def read(self, size=-1, *, timeout=None):
+        t = OptionalTimeout(timeout, getattr(self._console, 'timeout', None))
+
+        buf = b''
+
+        while not t.expired and (size < 0 or len(buf) < size):
+            # The timeout for read_nonblocking is per-character, so only read a
+            # single character at a time with a 1 second timeout so that we
+            # have timely checks for an expired timeout
+            try:
+                buf += self.read_nonblocking(1, t.remaining)
+            except TIMEOUT:
+                pass
+            except EOF:
+                break
+
+        return buf
+
+    def read_nonblocking(self, size=1, timeout=-1):
+        buf = b""
+
+        while size < 0 or len(buf) < size:
+            if not self._alive:
+                if not buf:
+                    raise EOF("ConsoleMarkerProcess end-of-file")
+                break
+
+            index, _, match, after = self._console.expect(
+                [self._output_re, self._eof_re, TIMEOUT],
+                timeout=timeout,
+            )
+            if index == 0:
+                buf = buf + after
+            elif index == 1:
+                self._handle_eof(match)
+                if not buf:
+                    raise EOF("ConsoleMarkerProcess end-of-file")
+                break
+            elif index == 2:
+                if not buf:
+                    raise TIMEOUT("ConsoleMarkerProcess timeout")
+                break
+
+        return buf
+
+
+    @step(args=["data"])
+    def write(self, data):
+        if self._alive:
+            return self._console.write(data)
+        return 0
+
+    @step(result=True)
+    def poll(self):
+        if not self._alive:
+            return self.exitcode
+
+        index, _, match, _ = self._console.expect([self._eof_re, TIMEOUT], timeout=1)
```

Comment:
This looks like it could drop output from the process. Is that intentional?

Reply:
I don't think it can because the EOF is anchored to the beginning of the string, so unless the output has all been consumed, the process will still be "alive" (see the warning in `wait`)